File: packages/butler-agent/src/agent/btcc/subsessions/service.ts
import { NativeInboundQueue } from "../../../gateways/core/inbound-queue.ts";
import { shortStewardWorktreeBranch } from "../../session-workspaces/index.ts";
import { digest } from "../identity/index.ts";
import { subsessionChildTurnId, subsessionDelegationId, subsessionResultId, subsessionRootWorkId } from "./identities.ts";
import { recoverPendingParentInputs } from "./outbox-recovery.ts";
import { completeStewardResultForDependencies } from "./terminal-result-service.ts";
import { resolveParentResultEvidence } from "./accepted-terminal-report.ts";
import { createStewardWorktree } from "./worktree.ts";
import { assertExactChildLedgerProjectIdentity, childProjectContextBinding, delegationProjectContextReady, snapshotChildProjectContext } from "./project-context.ts";
import {
  normalizeSubsessionAllowedToolsAndEffects,
  normalizeSubsessionMutationScopeForEffects,
} from "./scope.ts";
import { completePacketContext } from "./terminal-results.ts";
import { renderDelegatedParentConversationContext } from "./parent-conversation-context.ts";
import { renderStewardInput } from "./steward-input.ts";
import {
  assertReviewedDelegationRequest,
  loadReviewedDelegationPlan,
} from "./reviewed-delegation-plan.ts";
import {
  admittedParentTurnAccessMode,
  inheritedStewardRuntimePolicy,
  normalizeStewardAccessMode,
  reviewedStewardDelegationRequest,
  stewardRootWorkScope,
} from "./runtime-policy.ts";
import { createSubsessionControlService } from "./control.ts";
import type {
  CreatedDelegation, DelegationPacket, DelegationRequest, ParentInputSink,
  SessionRelation, SubsessionExecutionMode, SubsessionDelegationDependencies,
  SubsessionDelegationService,
} from "./contracts.ts";
import type { StoredSessionBinding } from "../../../test-support/harness/contracts.ts";
export function createSubsessionDelegationService(
  input: SubsessionDelegationDependencies,
): SubsessionDelegationService {
  const childQueue = new NativeInboundQueue(input.butlerData);
  const parentInputSink: ParentInputSink = input.parentInputSink;
  const completeResult = (resultInput: Parameters<SubsessionDelegationService["completeStewardResult"]>[0]) =>
    completeStewardResultForDependencies(input, parentInputSink, resultInput);
  const ensureRootWork = async (child: Parameters<SubsessionDelegationService["ensureChildRootWork"]>[0]): Promise<string> => {
    const relation = input.store.relationByChildSessionId(child.childSessionId);
    if (!relation) throw new Error("subsession_relation_missing");
    const expectedRootWorkId = input.store.rootWorkIdByRelationId(relation.relation_id);
    if (!expectedRootWorkId) throw new Error("subsession_root_work_identity_missing");
    const packet = input.store.packetByRelationId(relation.relation_id);
    if (!packet || !completePacketContext(packet) || !await delegationProjectContextReady(packet.project_context, { sessionId: child.childSessionId, turnId: child.childTurnId }, input)) {
      await completeResult({
        childSessionId: child.childSessionId,
        childTurnId: child.childTurnId,
        resultId: subsessionResultId(relation.child_session_id, child.childTurnId),
        status: "blocked",
        code: "delegation_context_incomplete",
      });
      throw new Error("delegation_context_incomplete");
    }
    const childBinding = input.sessionBindings.getBySessionId(child.childSessionId);
    if (!childBinding || childBinding.role !== "steward") {
      throw new Error("subsession_child_binding_missing");
    }
    assertExactChildLedgerProjectIdentity(childBinding);
    const rootWorkScope = stewardRootWorkScope(childBinding);
    const existing = await input.durableWork.boundWorkForTurn(child.childTurnId);
    if (existing) {
      if (existing.workId !== expectedRootWorkId || existing.sessionId !== child.childSessionId) {
        throw new Error("subsession_root_work_identity_mismatch");
      }
      return existing.workId;
    }
    const work = await input.durableWork.startWork({
      sessionId: child.childSessionId,
      turnId: child.childTurnId,
      ...rootWorkScope,
      mutationCallId: `subsession-root-work:${packet.delegation_id}:${packet.task_id}:${child.childSessionId}`,
      objective: child.objective,
    });
    if (work.workId !== expectedRootWorkId) throw new Error("subsession_root_work_identity_mismatch");
    return work.workId;
  };
  const service: SubsessionDelegationService = {
    async reviewedDelegationPlan(parentInput) {
      return loadReviewedDelegationPlan(input, parentInput);
    },
    async delegateReviewed(request) {
      const reviewed = await loadReviewedDelegationPlan(input, {
        parentSessionId: request.parent_session_id,
        parentTurnId: request.parent_turn_id,
      });
      return service.delegate(reviewedStewardDelegationRequest(request, reviewed));
    },
    async delegate(request) {
      const normalizedRequest = normalizeDelegationRequest(request);
      const parent = input.sessionBindings.getBySessionId(normalizedRequest.parent_session_id);
      if (!parent || parent.role !== "butler") throw new Error("parent_butler_session_required");
      const parentTurn = await input.parentTurns.findTurn(normalizedRequest.parent_turn_id);
      if (!parentTurn || parentTurn.sessionId !== normalizedRequest.parent_session_id) {
        throw new Error("subsession_parent_turn_required");
      }
      if (normalizedRequest.parent_access_mode !== admittedParentTurnAccessMode(parentTurn)) {
        throw new Error("subsession_parent_access_mode_mismatch");
      }
      if (normalizedRequest.model_ref !== parent.modelRef) throw new Error("subsession_parent_model_mismatch");
      const parentReasoning = parent.metadata?.reasoning_effort;
      if (typeof parentReasoning === "string" && parentReasoning !== normalizedRequest.reasoning_effort) throw new Error("subsession_parent_reasoning_mismatch");
      const { projectContext, inheritedProject } = await snapshotChildProjectContext({
        parentSessionId: normalizedRequest.parent_session_id, parentTurnId: normalizedRequest.parent_turn_id, parent,
        turns: input.parentTurns, documents: input.contextDocuments,
      });
      const reviewed = await loadReviewedDelegationPlan(input, {
        parentSessionId: normalizedRequest.parent_session_id,
        parentTurnId: normalizedRequest.parent_turn_id,
      });
      assertReviewedDelegationRequest(normalizedRequest, reviewed);
      const delegationId = subsessionDelegationId(normalizedRequest);
      const existing = input.store.relationByDelegationId(delegationId); if (existing) return recoverExistingDelegation(input, existing);
      const relationId = `relation-${digest(`btcc.subsession.relation.v1\0${delegationId}`).slice(0, 40)}`;
      // Persisted task_id is the managerial assignment, never a Worker Task.
      const managerialAssignmentId = `task-${digest(`btcc.subsession.task.v1\0${delegationId}`).slice(0, 40)}`;
      const childSessionId = `steward-${digest(`btcc.subsession.child-session.v1\0${relationId}`).slice(0, 32)}`;
      const childTurnId = `steward-turn-${digest(`btcc.subsession.child-turn.v1\0${relationId}`).slice(0, 32)}`;
      const rootWorkId = subsessionRootWorkId(delegationId, managerialAssignmentId, childSessionId);
      const branch = shortStewardWorktreeBranch(relationId);
      const parentChatId = parent.transportBindings.find((binding) =>
        binding.transport === "app" && binding.peerId.trim(),
      )?.peerId;
      if (!parentChatId) throw new Error("parent_app_binding_required");
      const now = new Date().toISOString();
      const relation: SessionRelation = {
        relation_id: relationId,
        parent_session_id: normalizedRequest.parent_session_id,
        parent_turn_id: normalizedRequest.parent_turn_id,
        child_session_id: childSessionId,
        anchor_message_id: normalizedRequest.anchor_message_id,
        ordinal: nextOrdinal(input, normalizedRequest.parent_session_id),
        safe_title: normalizedRequest.safe_title,
        created_at: now,
      };
      const packet = createPacket(normalizedRequest, {
        delegationId, relationId, taskId: managerialAssignmentId,
        parentWorkRef: reviewed.parent_work_ref,
        branch,
      }, projectContext);
      const parentConversationContext = renderDelegatedParentConversationContext({
        conversations: input.conversations,
        parentSessionId: normalizedRequest.parent_session_id,
        parentTurnId: normalizedRequest.parent_turn_id,
        modelRef: normalizedRequest.model_ref,
      });
      registerChildSession(input, parent, normalizedRequest, packet, childSessionId, inheritedProject);
      const childWorkspacePath = normalizedRequest.execution_mode === "read_only"
        ? parent.workspacePath
        : await createStewardWorktree(input, parent.workspacePath, branch, childSessionId);
      const storedChild = input.sessionBindings.getBySessionId(childSessionId);
      if (!storedChild || (normalizedRequest.execution_mode === "mutation" &&
        storedChild.workspacePath === parent.workspacePath)) {
        throw new Error("steward_isolated_workspace_missing");
      }
      input.store.create({ relation, packet, childTurnId, rootWorkId });
      enqueueChild(childQueue, packet, normalizedRequest.parent_session_id, childSessionId,
        childTurnId, childWorkspacePath, now, parentConversationContext);
      return { relation, packet, child_turn_id: childTurnId, root_work_id: rootWorkId,
        child_workspace_path: childWorkspacePath } satisfies CreatedDelegation;
    },
    async ensureChildRootWork(child) {
      return ensureRootWork(child);
    },
    async completeStewardResult(resultInput) {
      return completeResult(resultInput);
    },
    async recoverPendingParentInputs() {
      return recoverPendingParentInputs({ store: input.store, sink: parentInputSink });
    },
    async resolveParentResultEvidence(parentInput) {
      return resolveParentResultEvidence({ ...parentInput, store: input.store, turns: input.parentTurns });
    },
    resultIdForRelation(relationId) {
      return input.store.resultIdForRelation(relationId);
    },
    pendingParentInputCount() {
      return input.store.pendingParentInputCount();
    },
    ...createSubsessionControlService(input, childQueue),
  };
  return service;
}

function recoverExistingDelegation(
  input: SubsessionDelegationDependencies,
  relation: SessionRelation,
): CreatedDelegation {
  const packet = input.store.packetByRelationId(relation.relation_id);
  const child = input.sessionBindings.getBySessionId(relation.child_session_id);
  if (!packet || !child) throw new Error("subsession_existing_identity_incomplete");
  return {
    relation,
    packet,
    child_turn_id: subsessionChildTurnId(relation.relation_id),
    root_work_id: subsessionRootWorkId(packet.delegation_id, packet.task_id, relation.child_session_id),
    child_workspace_path: child.workspacePath,
  };
}
function createPacket(
  request: DelegationRequest,
  ids: { delegationId: string; relationId: string; taskId: string; parentWorkRef: DelegationPacket["parent_work_ref"]; branch: string },
  projectContext: DelegationPacket["project_context"],
): DelegationPacket {
  return {
    delegation_id: ids.delegationId,
    task_id: ids.taskId,
    parent_session_id: request.parent_session_id,
    parent_turn_id: request.parent_turn_id,
    relation_id: ids.relationId,
    execution_mode: request.execution_mode,
    objective: request.objective,
    acceptance_criteria: [...request.acceptance_criteria],
    task_or_plan_refs: [...request.task_or_plan_refs],
    ...(projectContext ? { project_context: projectContext } : {}),
    constraints_and_non_goals: [...request.constraints_and_non_goals],
    allowed_tools_and_effects: [...request.allowed_tools_and_effects],
    mutation_scope: [...request.mutation_scope],
    workspace_and_worktree: request.execution_mode === "read_only"
      ? {
          ownership: "project",
          workspace_label: "Validated project workspace",
          repository_anchor_ref: "parent-session-project",
        }
      : {
          ownership: "session",
          workspace_label: "Steward session worktree",
          repository_anchor_ref: "parent-session-repository",
          branch: ids.branch,
        },
    expected_result_schema: {
      version: 1,
      status: "success",
      required_fields: ["summary", "acceptance_evidence", "changed_artifacts"],
    },
    work_creation_policy: "one_recoverable_child_work",
    access_and_budget_policy: {
      access_mode: request.parent_access_mode,
      max_turns: 12,
      model_ref: request.model_ref,
      reasoning_effort: request.reasoning_effort,
    },
    parent_work_ref: ids.parentWorkRef,
    model_ref: request.model_ref,
    reasoning_effort: request.reasoning_effort,
  };
}
function registerChildSession(
  input: SubsessionDelegationDependencies,
  parent: StoredSessionBinding,
  request: DelegationRequest,
  packet: DelegationPacket,
  childSessionId: string,
  inheritedProject: ReturnType<typeof childProjectContextBinding>,
): void {
  input.sessionBindings.upsert({
    sessionId: childSessionId,
    role: "steward",
    ...(inheritedProject?.sessionBinding ?? {}),
    workspacePath: parent.workspacePath,
    runtimeAdapterId: "btcc-turn-runtime",
    modelProviderId: parent.modelProviderId,
    modelRef: packet.model_ref as `${string}/${string}`,
    transportBindings: [],
    metadata: {
      source: "btcc-subsession",
      subsession: {
        relation_id: packet.relation_id,
        delegation_id: packet.delegation_id,
        task_id: packet.task_id,
        parent_session_id: request.parent_session_id,
        execution_mode: packet.execution_mode,
        mutation_scope: [...packet.mutation_scope],
        allowed_tools_and_effects: [...packet.allowed_tools_and_effects],
        ...(inheritedProject ? { project_context: inheritedProject.metadata } : {}),
      },
      runtimePolicy: inheritedStewardRuntimePolicy(parent, request.parent_access_mode),
      reasoning_effort: packet.reasoning_effort,
    },
  });
}

function enqueueChild(
  childQueue: NativeInboundQueue,
  packet: DelegationPacket,
  parentSessionId: string,
  childSessionId: string,
  childTurnId: string,
  workspacePath: string,
  timestamp: string,
  parentConversationContext: string,
): void {
  childQueue.enqueueIdempotent({
    eventId: `steward:${packet.delegation_id}`,
    transport: "app",
    accountId: "local",
    peer: { kind: "dm", id: childSessionId, parentId: parentSessionId },
    sender: { id: "butler-steward-dispatch", displayName: "Butler Steward" },
    message: {
      id: `steward-message:${packet.delegation_id}`,
      text: renderStewardInput(packet, parentConversationContext),
      timestamp,
    },
    routingHints: { stewardId: childSessionId, turnId: childTurnId },
    nativeStewardContext: {
      version: 1,
      projectName: packet.project_context?.project_id ?? "",
      workspacePath,
      modelRef: packet.model_ref as `${string}/${string}`,
      reasoningEffort: packet.reasoning_effort,
    },
    raw: { source: "btcc-subsession-delegation" },
  });
}
function nextOrdinal(input: SubsessionDelegationDependencies, parentSessionId: string): number {
  return (input.store.relationsByParentSessionId(parentSessionId).at(-1)?.ordinal ?? 0) + 1;
}
function normalizeDelegationRequest(input: DelegationRequest): DelegationRequest {
  for (const [key, value] of Object.entries(input)) {
    if (typeof value === "string" && !value.trim()) throw new Error(`delegation_${key}_required`);
  }
  if (!input.parent_work_ref) throw new Error("subsession_parent_work_ref_required");
  const parentAccessMode = normalizeStewardAccessMode(input.parent_access_mode);
  const executionMode = normalizeExecutionMode(input.execution_mode);
  const allowedToolsAndEffects = normalizeSubsessionAllowedToolsAndEffects(
    input.allowed_tools_and_effects,
    executionMode,
  );
  const mutationScope = executionMode === "mutation"
    ? normalizeSubsessionMutationScopeForEffects(
        input.mutation_scope,
        allowedToolsAndEffects,
      )
    : [];
  return {
    ...input,
    parent_access_mode: parentAccessMode,
    execution_mode: executionMode,
    allowed_tools_and_effects: allowedToolsAndEffects,
    mutation_scope: mutationScope,
  };
}

function normalizeExecutionMode(value: unknown): SubsessionExecutionMode {
  if (value === "read_only" || value === "mutation") return value;
  throw new Error("delegation_execution_mode_invalid");
}

File: packages/butler-agent/rust/agent/src/btcc/subsessions/service.rs
//! Durable subsession orchestration and role-specific result routing.

mod control;
mod helpers;
mod projection;
mod result_delivery;
#[cfg(test)]
mod tests;
mod worker;

pub(crate) use control::{
    SubsessionCancelRequest, SubsessionDirectionRequest, SubsessionResumeRequest,
};

use helpers::*;

use serde_json::{Map, Value, json};
use std::sync::Arc;

use crate::btcc::{
    ActionStatus, BtccError, DurableWorkService, ExecutionMode, ParentResultRoute, PortFuture,
    SqliteSubsessionRepository, StartWorkInput, SubsessionCreate, WorkTurnScope, WorkView,
};
use crate::workspace::{OwnOptional, SessionBindingStore, SessionRole, UpsertSessionBinding};

#[derive(Clone, Debug)]
pub(crate) struct InterruptedSubsessionEvent {
    pub event_id: String,
    pub message_id: String,
    pub message: String,
    pub recovery_id: String,
}

#[derive(Clone, Debug)]
pub(crate) struct SubsessionEnqueue {
    pub envelope: Value,
    pub metadata: Map<String, Value>,
}

pub(crate) trait SubsessionChildQueue: Send + Sync {
    fn enqueue(&self, input: SubsessionEnqueue) -> Result<(), BtccError>;
    fn interrupted_event(
        &self,
        event_id: &str,
        session_id: &str,
        turn_id: &str,
    ) -> Result<Option<InterruptedSubsessionEvent>, BtccError>;
}

#[derive(Clone, Debug)]
pub(crate) struct WorkerProfile {
    pub id: String,
    pub label: String,
    pub enabled: bool,
    pub model_ref: String,
    pub reasoning_effort: String,
    pub prompt: Option<String>,
    pub job: Value,
}

pub(crate) trait WorkerProfileReader: Send + Sync {
    fn list(&self) -> PortFuture<'_, Vec<WorkerProfile>>;
    fn read(&self, profile_id: Option<String>) -> PortFuture<'_, WorkerProfile>;
}

#[derive(Clone, Debug)]
pub(crate) struct StewardDelegationRequest {
    pub parent_session_id: String,
    pub parent_turn_id: String,
    pub anchor_message_id: String,
    pub request: String,
    pub safe_title: Option<String>,
    pub model_ref: String,
    pub reasoning_effort: String,
    pub access_mode: String,
}

#[derive(Clone, Debug)]
pub(crate) struct WorkerDelegationRequest {
    pub parent_session_id: String,
    pub parent_turn_id: String,
    pub anchor_message_id: String,
    pub source_tool_call_id: String,
    pub action_key: String,
    pub objective: String,
    pub acceptance_criteria: Vec<String>,
    pub implementation_brief: String,
    pub safe_title: Option<String>,
    pub profile_id: Option<String>,
    pub access_mode: String,
}

#[derive(Clone)]
pub(crate) struct NativeSubsessionService {
    repository: SqliteSubsessionRepository,
    bindings: SessionBindingStore,
    queue: Arc<dyn SubsessionChildQueue>,
    profiles: Arc<dyn WorkerProfileReader>,
    work: Arc<DurableWorkService>,
    now: Arc<dyn Fn() -> String + Send + Sync>,
}

impl NativeSubsessionService {
    pub(crate) fn new(
        repository: SqliteSubsessionRepository,
        bindings: SessionBindingStore,
        queue: Arc<dyn SubsessionChildQueue>,
        profiles: Arc<dyn WorkerProfileReader>,
        work: Arc<DurableWorkService>,
        now: Arc<dyn Fn() -> String + Send + Sync>,
    ) -> Self {
        Self {
            repository,
            bindings,
            queue,
            profiles,
            work,
            now,
        }
    }

    pub(crate) async fn delegate_steward(
        &self,
        request: StewardDelegationRequest,
        reviewed: &WorkView,
    ) -> Result<Value, BtccError> {
        let parent = self
            .bindings
            .get_by_session_id(&request.parent_session_id)
            .await
            .map_err(workspace)?
            .ok_or_else(|| error("parent_butler_session_required"))?;
        if parent.role != SessionRole::Butler {
            return Err(error("parent_butler_session_required"));
        }
        let plan = reviewed
            .current_plan
            .as_ref()
            .ok_or_else(|| error("delegation_reviewed_plan_required"))?;
        if plan.execution_mode != Some(ExecutionMode::Steward) {
            return Err(error("steward_delegation_plan_mode_required"));
        }
        let review = reviewed
            .latest_plan_review
            .as_ref()
            .filter(|r| {
                r.verdict == crate::btcc::ReviewVerdict::Accept
                    && r.bound_plan_revision_id.as_deref() == Some(&plan.plan_revision_id)
            })
            .ok_or_else(|| error("delegation_reviewed_plan_required"))?;
        let parent_chat_id = parent
            .transport_bindings
            .iter()
            .find(|binding| binding.transport == "app" && !binding.peer_id.trim().is_empty())
            .map(|binding| binding.peer_id.clone())
            .ok_or_else(|| error("parent_app_binding_required"))?;
        let identity = json!({"parent_session_id":request.parent_session_id,"parent_turn_id":request.parent_turn_id,"request":request.request,"work_id":reviewed.work_id,"plan_revision_id":plan.plan_revision_id,"review_revision_id":review.review_revision_id});
        let encoded =
            serde_json::to_string(&identity).map_err(|_| error("subsession_identity_invalid"))?;
        let delegation_id = format!(
            "delegation-{}",
            crate::btcc::digest_identity(&format!("btcc.subsession.delegation.v1\0{encoded}"))
        );
        if let Some(existing) = self
            .repository
            .by_delegation(delegation_id.clone())
            .await
            .map_err(storage)?
        {
            self.ensure_child_binding(&existing).await?;
            self.replay(&existing).await?;
            return Ok(delegation_output(&existing));
        }
        let relation_id = format!(
            "relation-{}",
            &crate::btcc::digest_identity(&format!("btcc.subsession.relation.v1\0{delegation_id}"))
                [..40]
        );
        let task_id = format!(
            "task-{}",
            &crate::btcc::digest_identity(&format!("btcc.subsession.task.v1\0{delegation_id}"))
                [..40]
        );
        let child_session_id = format!(
            "steward-{}",
            &crate::btcc::digest_identity(&format!(
                "btcc.subsession.child-session.v1\0{relation_id}"
            ))[..32]
        );
        let child_turn_id = format!(
            "steward-turn-{}",
            &crate::btcc::digest_identity(&format!("btcc.subsession.child-turn.v1\0{relation_id}"))
                [..32]
        );
        let mutation_call_id =
            format!("subsession-root-work:{delegation_id}:{task_id}:{child_session_id}");
        let root_work_id = format!(
            "guided-work-{}",
            crate::btcc::digest_identity(&format!("btcc-guided-work.v1\0work\0{mutation_call_id}"))
        );
        let now = (self.now)();
        let packet = json!({"child_role":"steward","delegation_id":delegation_id,"task_id":task_id,"parent_session_id":request.parent_session_id,"parent_turn_id":request.parent_turn_id,"parent_chat_id":parent_chat_id,"relation_id":relation_id,"access_mode":request.access_mode,"execution_mode":if request.access_mode=="read_only"{"read_only"}else{"mutation"},"objective":request.request,"acceptance_criteria":plan.checks,"task_or_plan_refs":[plan.plan_revision_id],"constraints_and_non_goals":[],"allowed_tools_and_effects":allowed_effects(&request.access_mode),"mutation_scope":mutation_scope(&request.access_mode),"parent_work_ref":{"work_id":reviewed.work_id,"session_id":reviewed.session_id,"turn_id":request.parent_turn_id,"plan_revision_id":plan.plan_revision_id,"review_revision_id":review.review_revision_id},"model_ref":request.model_ref,"reasoning_effort":request.reasoning_effort});
        let envelope = child_envelope(ChildEnvelopeInput {
            role: "steward",
            delegation: &delegation_id,
            child: &child_session_id,
            parent_id: &request.parent_session_id,
            turn: &child_turn_id,
            parent: &parent,
            model: &request.model_ref,
            reasoning: &request.reasoning_effort,
            text: render_steward_input(&packet),
            now: &now,
        });
        let intent =
            json!({"envelope":envelope,"metadata":{"source":"btcc-subsession-delegation"}});
        self.repository
            .create(SubsessionCreate {
                relation_id: relation_id.clone(),
                delegation_id: delegation_id.clone(),
                task_id,
                parent_session_id: request.parent_session_id,
                parent_turn_id: request.parent_turn_id,
                child_session_id: child_session_id.clone(),
                child_turn_id,
                anchor_message_id: request.anchor_message_id,
                safe_title: request
                    .safe_title
                    .unwrap_or_else(|| "Delegated task".into()),
                root_work_id,
                packet,
                dispatch_intent: intent,
                created_at: now,
            })
            .await
            .map_err(storage)?;
        let stored = self
            .repository
            .by_delegation(delegation_id)
            .await
            .map_err(storage)?
            .ok_or_else(|| error("subsession_persist_failed"))?;
        self.ensure_child_binding(&stored).await?;
        self.replay(&stored).await?;
        Ok(delegation_output(&stored))
    }

    async fn create_child_binding(
        &self,
        stored: &crate::btcc::StoredSubsessionDelegation,
        parent: &crate::workspace::StoredSessionBinding,
        role: SessionRole,
        metadata: Value,
    ) -> Result<(), BtccError> {
        self.bindings
            .upsert(UpsertSessionBinding {
                session_id: stored.child_session_id.clone(),
                role,
                project_id: parent.project_id.clone(),
                app_project_id: parent
                    .app_project_id
                    .clone()
                    .map_or(OwnOptional::Null, OwnOptional::Value),
                ledger_project_id: parent
                    .ledger_project_id
                    .clone()
                    .map_or(OwnOptional::Null, OwnOptional::Value),
                workspace_path: parent.workspace_path.clone(),
                runtime_adapter_id: "btcc-turn-runtime".into(),
                model_provider_id: stored.packet["model_ref"]
                    .as_str()
                    .unwrap_or("")
                    .split('/')
                    .next()
                    .unwrap_or("")
                    .into(),
                model_ref: stored.packet["model_ref"].as_str().unwrap_or("").into(),
                runtime_session_ref: None,
                provider_thread_ref: None,
                transport_bindings: Vec::new(),
                lifecycle_state: None,
                created_at: None,
                updated_at: None,
                last_active_at: None,
                metadata: metadata.as_object().cloned(),
            })
            .await
            .map_err(workspace)?;
        Ok(())
    }
    async fn ensure_child_binding(
        &self,
        stored: &crate::btcc::StoredSubsessionDelegation,
    ) -> Result<(), BtccError> {
        let (role, role_name) = match stored.packet["child_role"].as_str() {
            Some("steward") => (SessionRole::Steward, "steward"),
            Some("worker") => (SessionRole::Worker, "worker"),
            _ => return Err(error("subsession_child_role_invalid")),
        };
        if let Some(existing) = self
            .bindings
            .get_by_session_id(&stored.child_session_id)
            .await
            .map_err(workspace)?
        {
            if existing.role != role {
                return Err(error("subsession_child_binding_mismatch"));
            }
            return Ok(());
        }
        let parent = self
            .bindings
            .get_by_session_id(&stored.parent_session_id)
            .await
            .map_err(workspace)?
            .ok_or_else(|| error("subsession_parent_binding_missing"))?;
        let access = stored.packet["access_mode"]
            .as_str()
            .ok_or_else(|| error("subsession_access_mode_invalid"))?;
        let metadata = child_metadata(role_name, &stored.packet, access, &parent);
        self.create_child_binding(stored, &parent, role, metadata)
            .await
    }
    async fn replay(
        &self,
        stored: &crate::btcc::StoredSubsessionDelegation,
    ) -> Result<(), BtccError> {
        let envelope = stored
            .dispatch_intent
            .get("envelope")
            .cloned()
            .ok_or_else(|| error("subsession_dispatch_intent_invalid"))?;
        self.queue.enqueue(SubsessionEnqueue {
            envelope,
            metadata: Map::new(),
        })?;
        self.repository
            .mark_enqueued(stored.relation_id.clone())
            .await
            .map_err(storage)
    }
    pub(crate) async fn ensure_child_work(
        &self,
        session: &str,
        turn: &str,
    ) -> Result<(), BtccError> {
        let stored = self
            .repository
            .by_child(session.into())
            .await
            .map_err(storage)?
            .ok_or_else(|| error("subsession_relation_missing"))?;
        let binding = self
            .bindings
            .get_by_session_id(session)
            .await
            .map_err(workspace)?
            .ok_or_else(|| error("subsession_child_binding_missing"))?;
        let scope = child_work_scope(&stored, &binding, turn)?;
        let existing = self.work.bound_work_for_turn(turn.into()).await?;
        if let Some(existing) = existing {
            if existing.work_id != stored.root_work_id
                || existing.session_id != session
                || !work_matches_scope(&existing, &scope)
            {
                return Err(error("subsession_root_work_identity_mismatch"));
            }
            return Ok(());
        }
        if turn != stored.child_turn_id {
            let bound = self
                .work
                .bind_open_work(scope.clone(), Some(stored.root_work_id.clone()))
                .await?
                .ok_or_else(|| error("subsession_root_work_identity_mismatch"))?;
            if bound.work_id != stored.root_work_id
                || bound.session_id != session
                || !work_matches_scope(&bound, &scope)
            {
                return Err(error("subsession_root_work_identity_mismatch"));
            }
            return Ok(());
        }
        let mutation = format!(
            "subsession-root-work:{}:{}:{}",
            stored.delegation_id, stored.task_id, stored.child_session_id
        );
        let work = self
            .work
            .start_work(StartWorkInput {
                scope,
                mutation_call_id: mutation,
                objective: stored.packet["objective"]
                    .as_str()
                    .unwrap_or("Complete the bounded subsession task.")
                    .into(),
                backfill_tool_call_ids: None,
            })
            .await?;
        if work.work_id != stored.root_work_id
            || work.session_id != session
            || !work_matches_scope(&work, &child_work_scope(&stored, &binding, turn)?)
        {
            return Err(error("subsession_root_work_identity_mismatch"));
        }
        Ok(())
    }
    pub(crate) async fn complete_child(
        &self,
        session: &str,
        turn: &str,
        status: &str,
        summary: String,
    ) -> Result<(), BtccError> {
        self.ensure_child_work(session, turn).await?;
        if status == "cancelled" {
            self.work
                .abandon_bound_work_for_turn(turn.to_owned())
                .await?;
        }
        let mut evidence_refs = self
            .work
            .bound_work_for_turn(turn.to_owned())
            .await?
            .and_then(|work| work.latest_disposition)
            .map(|disposition| disposition.evidence_snapshot)
            .unwrap_or_default();
        evidence_refs.extend(
            self.repository
                .child_result_evidence(session.to_owned())
                .await
                .map_err(storage)?,
        );
        evidence_refs.sort();
        evidence_refs.dedup();
        self.repository
            .commit_result(
                session.into(),
                turn.into(),
                status.into(),
                summary,
                evidence_refs,
                (self.now)(),
            )
            .await
            .map_err(storage)?;
        self.deliver_worker_results().await
    }
    pub(crate) async fn recover_dispatches(&self) -> Result<(), BtccError> {
        for stored in self
            .repository
            .pending_dispatches()
            .await
            .map_err(storage)?
        {
            self.ensure_child_binding(&stored).await?;
            self.replay(&stored).await?;
        }
        self.recover_directions().await?;
        self.deliver_worker_results().await
    }
    pub(crate) async fn should_wait_for_child(&self, parent: &str) -> Result<bool, BtccError> {
        self.repository
            .has_active_child(parent.to_owned())
            .await
            .map_err(storage)
    }
    pub(crate) async fn has_unfinished_execution(
        &self,
        session_id: &str,
    ) -> Result<bool, BtccError> {
        self.repository
            .has_unfinished_execution(session_id.to_owned())
            .await
            .map_err(storage)
    }
    pub(crate) fn repository(&self) -> SqliteSubsessionRepository {
        self.repository.clone()
    }
    pub(crate) async fn enabled_worker_profiles(&self) -> Result<Vec<WorkerProfile>, BtccError> {
        Ok(self
            .profiles
            .list()
            .await?
            .into_iter()
            .filter(|profile| profile.enabled)
            .collect())
    }
}
